### PR TITLE
Support IsIndexer for ViewComponent tag helpers

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - #5207 

@NTaylorMullen

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/Internal/ViewComponentTagHelperChunkVisitor.cs [hunk 1]
```diff
@@ -131,9 +131,23 @@ private void BuildAttributeDeclarations(TagHelperDescriptor descriptor)
...
             foreach (var attribute in descriptor.Attributes)
             {
+                if (attribute.IsIndexer)
+                {
+                    continue;
+                }
+
                 Writer.WriteAutoPropertyDeclaration("public", attribute.TypeName, attribute.PropertyName);
+
+                if (indexerAttributes.Any(a => a.PropertyName == attribute.PropertyName))
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/Internal/ViewComponentTagHelperChunkVisitor.cs]: string.Equals

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/Internal/ViewComponentTagHelperChunkVisitor.cs [hunk 2]
```diff
@@ -131,9 +131,23 @@ private void BuildAttributeDeclarations(TagHelperDescriptor descriptor)
...
+                {
+                    Writer.Write(" = ")
+                        .WriteStartNewObject(attribute.TypeName)
+                        .WriteEndMethodInvocation(endLine: true);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/Internal/ViewComponentTagHelperChunkVisitor.cs]: `endLine: true` is the default

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewComponentTagHelperDescriptorFactory.cs [hunk 3]
```diff
@@ -105,14 +108,52 @@ private TagHelperDescriptor CreateDescriptor(ViewComponentDescriptor viewCompone
                 descriptor.IsIndexer = false;
 
                 attributeDescriptors.Add(descriptor);
+
+                var indexerDescriptor = GetIndexerAttributeDescriptor(parameter, lowerKebabName);
+                if (indexerDescriptor != null)
+                {
+                    indexerDescriptors.Add(indexerDescriptor);
+                }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewComponentTagHelperDescriptorFactory.cs]: Add a comment stating why we don't add required attribute descriptors for indexers

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewComponentTagHelperDescriptorFactory.cs [hunk 4]
```diff
@@ -105,14 +108,52 @@ private TagHelperDescriptor CreateDescriptor(ViewComponentDescriptor viewCompone
...
+
+            var type = dictionaryTypeArguments[1];
+            var descriptor = new TagHelperAttributeDescriptor
+            {
+                Name = name + "-",
+                PropertyName = parameter.Name,
+                TypeName = GetCSharpTypeName(type)
+            };
+
+            descriptor.IsEnum = type.GetTypeInfo().IsEnum;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewComponentTagHelperDescriptorFactory.cs]: Set these above.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Internal/ChunkVisitorTestFactory.cs [hunk 5]
```diff
@@ -44,6 +44,7 @@ public static IList<Chunk> GetTestChunks(bool visitedTagHelperChunks)
                 GetTagHelperChunk("Baz"),
                 GetNestedViewComponentTagHelperChunk("Foo", visitedTagHelperChunks),
                 GetViewComponentTagHelperChunk("Bar", visitedTagHelperChunks),
+                GetIndexerViewComponentTagHelperChunk("Bee", visitedTagHelperChunks, "Dictionary<string, List<int>>"),
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Internal/ChunkVisitorTestFactory.cs]: Just so the generated output is correct, update the attribute type name to be the full name representation.

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/Internal/ViewComponentTagHelperChunkVisitor.cs [hunk 6]
```diff
@@ -131,9 +131,23 @@ private void BuildAttributeDeclarations(TagHelperDescriptor descriptor)
...
+            var indexerAttributes = descriptor.Attributes.Where(a => a.IsIndexer);
+
             foreach (var attribute in descriptor.Attributes)
             {
+                if (attribute.IsIndexer)
+                {
+                    continue;
+                }
+
                 Writer.WriteAutoPropertyDeclaration("public", attribute.TypeName, attribute.PropertyName);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/Internal/ViewComponentTagHelperChunkVisitor.cs]: Should totally add an overload to this that lets you specify end line or not. Another day though.

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/RazorPageActivatorTest.cs [hunk 7]
```diff
@@ -66,7 +66,7 @@ public void Activate_ActivatesAndContextualizesPropertiesOnViews()
             var viewContext = new ViewContext(
                 actionContext,
                 Mock.Of<IView>(),
-                new ViewDataDictionary(new EmptyModelMetadataProvider()),
+                new ViewDataDictionary(new EmptyModelMetadataProvider(), new ModelStateDictionary()),
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/RazorPageActivatorTest.cs]: /cc @jbagga, Ajay took care of some of this for you to let him use shared sources.

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewComponentTagHelperDescriptorFactory.cs (hunk 3 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewComponentTagHelperDescriptorFactory.cs]: Added.

**Comment by ajaybhargavb** [comment]
🆙 📅

**Comment by ajaybhargavb** [comment]
🆙 📅 for real

**Comment by NTaylorMullen** [comment]
:shipit: once you fix the comment

**Pull request merged**